src/combinatorics/simple_graph/fin_subgraph.lean

```diff
@@ -0,0 +1,119 @@
+Authors: Joanna Choules
+-/
+import topology.category.Top.limits
+import combinatorics.simple_graph.subgraph
+
+/-!
+# Homomorphisms from finite subgraphs
+
+This file defines the type of finite subgraphs of a `simple_graph` and proves a compactness result
+for homomorphisms to a finite codomain.
+
+## Main statements
+
+* `simple_graph.exists_hom_of_all_finite_homs`: If every finite subgraph of a (possibly infinite)
+  graph `G` has a homomorphism to some finite graph `F`, then there is also a homomorphism `G →g F`.
+
+## Notations
+
+`→fg` is a variant on `→g` where the domain is a finite subgraph of some supergraph `G`.
+
+## Implementation notes
+
+The proof here uses compactness as formulated in `nonempty_sections_of_fintype_inverse_system`. For
+finite subgraphs `G'' ≤ G'`, the inverse system `fin_subgraph_hom_functor` restricts homomorphisms
+`G' →fg F` to domain `G''`.
+-/
+
+universes u v
+variables {V : Type u} {W : Type v} {G : simple_graph V} {F : simple_graph W}
+
+/-- The subtype of `G.subgraph` comprising those subgraphs with finite vertex sets. -/
+abbreviation simple_graph.fin_subgraph (G : simple_graph V) :=
+{ G' : G.subgraph // G'.verts.finite }
+
+/-- A graph homomorphism from a finite subgraph of G to F. -/
+abbreviation fin_subgraph_hom (G' : G.fin_subgraph) (F : simple_graph W) := G'.val.coe →g F
+
+local infix ` →fg ` : 50 := fin_subgraph_hom
+
+/-- The finite subgraph of G generated by a single vertex. -/
+def vertex_singleton (v : V) : G.fin_subgraph := ⟨simple_graph.singleton_subgraph _ v, by simp⟩
+
+/-- The finite subgraph of G generated by a single edge. -/
+def edge_singleton {u v : V} (e : G.adj u v) : G.fin_subgraph :=
+⟨simple_graph.subgraph_of_adj _ e, by simp⟩
+
+/- Lemmas establishing the ordering between edge- and vertex-generated subgraphs. -/
+
+lemma vertex_left_le_edge {u v : V} {e : G.adj u v} : vertex_singleton u ≤ edge_singleton e :=
+begin
+  unfold vertex_singleton edge_singleton, simp,
+end
+
+lemma vertex_right_le_edge {u v : V} {e : G.adj u v} : vertex_singleton v ≤ edge_singleton e :=
+begin
+  unfold vertex_singleton edge_singleton, simp,
+end
+
+/-- Given a homomorphism from a subgraph to `F`, construct its restriction to a sub-subgraph. -/
+def fin_subgraph_hom_restrict {G' G'' : G.fin_subgraph} (h : G'' ≤ G') (f : G' →fg F) : G'' →fg F :=
+begin
+  refine ⟨λ ⟨v, hv⟩, f.to_fun ⟨v, h.1 hv⟩, _⟩,
+  rintros ⟨u, hu⟩ ⟨v, hv⟩ huv,
+  exact f.map_rel' (h.2 huv),
+end
+
+/-- The inverse system of finite homomorphisms. -/
+def fin_subgraph_hom_functor (G : simple_graph V) (F : simple_graph W) :
+  (G.fin_subgraph)ᵒᵖ ⥤ Type (max u v) :=
+{ obj := λ G', G'.unop →fg F,
+  map := λ G' G'' g f, fin_subgraph_hom_restrict (category_theory.le_of_hom g.unop) f, }
+
+/-- If every finite subgraph of a graph `G` has a homomorphism to a finite graph `F`, then there is
+a homomorphism from the whole of `G` to `F`. -/
+lemma simple_graph.exists_hom_of_all_finite_homs [finite W]
```

Comment:
This should still hold true, but it might strongly break your proof.
```suggestion
lemma simple_graph.exists_hom_of_all_finite_homs [countable W]
```

Reply:
This breaks the proof because it makes the theorem false :wink:

Having the codomain be finite is essential, and there's no generalization to arbitrary `complete_graph α`. Though the purpose of this theorem is indeed that it can be applied to `α`-colorings. For example, lifting a proof of the 4-color theorem for finite planar graphs to infinite planar graphs.

I'll leave finding counterexamples as an exercise, but let me know if you want one!

Reply:
Aaah sorry I got my variables mixed up and thought `W` was the domain.